todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/view/TasksViews.kt

```diff
@@ -0,0 +1,167 @@
+        showMessage(R.string.completed_tasks_cleared)
+    }
+
+    override fun showLoadingTasksError() {
+        showMessage(R.string.loading_tasks_error)
+    }
+
+    private fun showMessage(messageRes: Int) {
+        Snackbar.make(rootView, messageRes, Snackbar.LENGTH_LONG).show()
+    }
+
+    override fun connect(output: Consumer<TasksListEvent>): Connection<TasksListViewData> {
+
+        addUiListeners(output)
+        val disposable = menuEvents.subscribe({ output.accept(it) }, {})
+
+        return onAccept<TasksListViewData> { render(it) }
+                .onDispose {
+                    disposable.dispose()
+                    mNoTaskAddView.setOnClickListener(null)
+                    mFab.setOnClickListener(null)
+                    mSwipeRefreshLayout.setOnRefreshListener(null)
+                    mListAdapter.setItemListener(null)
+                }
+    }
+
+    private fun addUiListeners(output: Consumer<TasksListEvent>) {
+        mNoTaskAddView.setOnClickListener { output.accept(NewTaskClicked) }
+        mFab.setOnClickListener { output.accept(NewTaskClicked) }
+        mSwipeRefreshLayout.setOnRefreshListener { output.accept(RefreshRequested) }
+        mListAdapter.setItemListener(
+                object : TasksAdapter.TaskItemListener {
+                    override fun onTaskClick(id: String) {
+                        output.accept(NavigateToTaskDetailsRequested(id))
+                    }
+
+                    override fun onCompleteTaskClick(id: String) {
+                        output.accept(TaskMarkedComplete(id))
+                    }
+
+                    override fun onActivateTaskClick(id: String) {
+                        output.accept(TaskMarkedActive(id))
+                    }
+                })
+    }
+
+    private fun showEmptyTaskState(vd: EmptyTasksViewData) {
+        mTasksView.visibility = View.GONE
+        mNoTasksView.visibility = View.VISIBLE
+
+        mNoTaskMainView.setText(vd.title)
+        mNoTaskIcon.setImageResource(vd.icon)
+        mNoTaskAddView.visibility = vd.addButtonVisibility
+    }
+
+    private fun showNoTasksViewState() {
+        mTasksView.visibility = View.GONE
+        mNoTasksView.visibility = View.GONE
+    }
+
+    private fun showTasks(tasks: List<TaskViewData>) {
+        mListAdapter.replaceData(tasks)
+
+        mTasksView.visibility = View.VISIBLE
+        mNoTasksView.visibility = View.GONE
+    }
+
+    private fun render(value: TasksListViewData) {
+        // Make sure setRefreshing() is called after the layout is done with everything else.
+        mSwipeRefreshLayout.isRefreshing = value.loading
+        mFilteringLabelView.setText(value.filterLabel)
+        when(value.viewState) {
+            is HasTasks -> showTasks(value.viewState.tasks)
+            AwaitingTasks -> showNoTasksViewState()
+            is EmptyTasks -> showEmptyTaskState(value.viewState.viewData)
+            else -> throw RuntimeException("Unhandled case")
```

Comment:
ViewState is a sealed class, so there should be no need for an else case

Reply:
But when isn't exhaustive here, which is why I added the else and kept it even though the IDE suggested I remove it

Reply:
Which case is missing?

Reply:
ooh do you mean that it's not used as an expression so it won't check that it's exhaustive?

Reply:
yup.. this is protection from future me adding more cases and forgetting to handle them :)

Reply:
Maybe the `.let {}` thing we discussed would be better in that case so the compiler will catch it?

This shows the `.exhaustive` thing I was talking about by the way :) https://proandroiddev.com/til-when-is-when-exhaustive-31d69f630a8b